numba_typing/test_overload_list.py

```diff
@@ -0,0 +1,146 @@
+    test = f"""
+def test_myfunc_{func_name}_type_default(self):
+    def foo({value_keys},{defaults_str}):
+        ...
+
+    @overload_list.overload_list(foo)
+    def foo_ovld_list():
+
+        def foo_{func_name}({value_str},{defaults_str_type}):
+            return ("{value_type}","{defaults_type}")
+
+        return (foo_{func_name},)
+
+    @njit
+    def jit_func({value_keys},{defaults_str}):
+        return foo({value_keys},{defaults_keys})
+
+    self.assertEqual(jit_func({param_qwe}), ("{value_type}", "{defaults_type}"))
+"""
+    loc = {}
+    exec(test, globals(), loc)
+    return loc
+
+
+L = List([1, 2, 3])
+L_int = List([List([1, 2])])
+L_float = List([List([List([3.0, 4.0])])])
+L_f = List([1.0, 2.0])
+D = Dict.empty(key_type=types.unicode_type, value_type=types.int64)
+D_1 = Dict.empty(key_type=types.int64, value_type=types.boolean)
+D['qwe'] = 1
+D['qaz'] = 2
+D_1[1] = True
+D_1[0] = False
+list_type = types.ListType(types.int64)
+D_list = Dict.empty(key_type=types.unicode_type, value_type=list_type)
+D_list['qwe'] = List([3, 4, 5])
+str_1 = 'qwe'
+str_2 = 'qaz'
+test_cases = [('int', [1, 2], {'a': int, 'b': int}), ('float', [1.0, 2.0], {'a': float, 'b': float}),
+              ('bool', [True, True], {'a': bool, 'b': bool}), ('str', ['str_1', 'str_2'], {'a': str, 'b': str}),
+              ('list', [[1, 2], [3, 4]], {'a': typing.List[int], 'b':list}),
+              ('List_typed', [L, [3, 4]], {'a': typing.List[int], 'b':list}),
+              ('tuple', [(1, 2.0), ('3', False)], {'a': typing.Tuple[int, float], 'b':tuple}),
+              ('dict', ['D', 'D_1'], {'a': typing.Dict[str, int], 'b': typing.Dict[int, bool]}),
+              ('union_1', [1, False], {'a': typing.Union[int, str], 'b': typing.Union[float, bool]}),
+              ('union_2', ['str_1', False], {'a': typing.Union[int, str], 'b': typing.Union[float, bool]}),
+              ('nested_list', ['L_int', 'L_float'], {'a': typing.List[typing.List[int]],
+                                                     'b': typing.List[typing.List[typing.List[float]]]}),
+              ('TypeVar_TT', ['L_f', [3.0, 4.0]], {'a': 'T', 'b': 'T'}),
+              ('TypeVar_TK', [1.0, 2], {'a': 'T', 'b': 'K'}),
+              ('TypeVar_ListT_T', ['L', 5], {'a': 'typing.List[T]', 'b': 'T'}),
+              ('TypeVar_ListT_DictKT', ['L', 'D'], {'a': 'typing.List[T]', 'b': 'typing.Dict[K, T]'}),
+              ('TypeVar_ListT_DictK_ListT', ['L', 'D_list'], {'a': 'typing.List[T]',
+                                                              'b': 'typing.Dict[K, typing.List[T]]'})]
+
+test_cases_default = [('int_defaults', [1], {'a': int}, {'b': 0}), ('float_defaults', [1.0], {'a': float}, {'b': 0.0}),
+                      ('bool_defaults', [True], {'a': bool}, {'b': False}),
+                      ('str_defaults', ['str_1'], {'a': str}, {'b': '0'}),
+                      ('tuple_defaults', [(1, 2)], {'a': tuple}, {'b': (0, 0)})]
+
+
+for name, val, annotation in test_cases:
```

Comment:
I'd prefer you to group all cases into 4-5 tests with subtests. Something like this:

```python
def test_common_types():
    test_cases = [({'a': 0}, {'a': int}),
                 ({'a': 0.}, {'a': float}),
                ...]
    
    for case in test_cases:
        with self.Subtest(case=case):
            run_test(case)
```